cirq-core/cirq/transformers/convert_to_target_gateset.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Callable, TYPE_CHECKING
+
+from cirq import protocols
+from cirq.transformers import transformer_api, transformer_primitives
+from cirq.protocols.decompose_protocol import DecomposeResult
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _create_on_stuck_raise_error(gateset: 'cirq.Gateset'):
+    def _value_error_describing_bad_operation(op: 'cirq.Operation') -> ValueError:
+        return ValueError(f"Unable to convert {op} to target gateset {gateset!r}")
+
+    return _value_error_describing_bad_operation
+
+
+@transformer_api.transformer
+def decompose_operations_to_target_gateset(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    gateset: Optional['cirq.Gateset'] = None,
+    decomposer: Callable[['cirq.Operation', int], DecomposeResult] = lambda *_: NotImplemented,
+    ignore_failures=True,
+) -> 'cirq.Circuit':
```

Comment:
Do we need this transformer ? It seems like a user should be able to fairly easily accomplish this same thing by just calling into the regular `decompose`. Having this as a thin wrapper over decompose seems like it might be complicating things for the users.

Reply:
Replied in the main comment. Tl;Dr is that having benefits of making it a transformer, like automated logging + no-compile tags, make it worth it. Plus we have other similar wrappers like `expand_composite`.